### ISSUE TITLE
Explicit API to begin swing-store transaction

### ISSUE THREAD

**Issue by mhofman** [Agoric/Agoric#8435]
## What is the Problem Being Solved?

As described in https://github.com/Agoric/agoric-sdk/issues/8423#issuecomment-1747286335, an explicit API for the host to begin a block transaction would likely have helped us uncover the missing `ensureTx`.

While #8432 would effectively create a transaction before any block work is started, that transaction would still be implicitly started by swing-store. We should consider an API for explicitly starting a swing-store transaction.

## Description of the Design

- New `startTransaction` which issues the `BEGIN IMMEDIATE TRANSACTION`, and sets a flag "inTransaction"
    - If already in a transaction, fail
- change `ensureTx` to check that `inTransaction` is true.
- update `commit` to reset the `inTransaction`

One complication of the non idempotency of this API for the host is that it sometimes needs to start a transaction before "begin block" for example in the case of upgrades or bootstrap, yet we want to maintain a single commit point at the end of the block.

## Security Considerations

None

## Scaling Considerations

None

## Test Plan

TBD
Make sure upgrades that effect swing-store changes run without issue

## Upgrade Considerations

Besides testing, none.

**Comment by FUDCo** [comment]
I think this needs a clearer statement of when it would be used.  Exposing this piece of swingstore semantics to the host makes me nervous, because part of the purpose of the swingstore API is to protect the host developer from having to have a model in mind of what's going on inside the swingstore as it operates.

**Comment by FUDCo** [comment]
On the other hand, we _do_ leave the decision as to when to `commit` up to the host, so perhaps this just introduces symmetry to the API.

**Comment by FUDCo** [comment]
Part of the problem that we're grappling with is that historically we've had a pretty clear sense of the where the boundary is when work is done, but the boundary of when work starts has always been a bit blurry.

**Comment by mhofman** [comment]
Exactly. The host is already in charge of the commit semantics. This just puts it in charge of what is meant to be committed. There are definitely overlaps however with #8089, and possibly another argument for "start migration" to be an explicit API.

**Comment by mhofman** [comment]
> Part of the problem that we're grappling with is that historically we've had a pretty clear sense of the where the boundary is when work is done, but the boundary of when work starts has always been a bit blurry.

The boundary is fairly clear from the host's perspective. But I guess it does conflict with some of the implicit creation and implicit upgrades semantics swing-store currently has or is considering.

**Comment by JimLarson** [comment]
Consider creating a complete wrapper around the SQL-lite connection object so that we can prevent any implicit creation of transactions.

**Comment by mhofman** [comment]
> Consider creating a complete wrapper around the SQL-lite connection object so that we can prevent any implicit creation of transactions.

That feels a bit heavy handed and not worth the complexity.
